### PR TITLE
HiSparse for Sparse Attention

### PR DESCRIPTION
**Pull request by xiezhq-hermann**
## Motivation
This PR introduces HiSparse, which leverages CPU memory to store idle KV cache during decoding, thereby increasing batch size and improving throughput for models that use the NSA sparse attention mechanism, such as DeepSeek-V3.2 and GLM-5. 
This PR is followed by a prior attempt to support hicache for sparse models: https://github.com/sgl-project/sglang/pull/14619, contributed by @hzh0425 @huangtingwei9988 as well.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/model_executor/model_runner_kv_cache_mixin.py [hunk 1]
```diff
@@ -460,6 +464,7 @@ def _init_pools(self: ModelRunner):
                     enable_memory_saver=self.server_args.enable_memory_saver,
                     start_layer=self.start_layer,
                     end_layer=self.end_layer,
+                    enable_hisparse=self.enable_hisparse,
```
Comment (alphabetc1) [inline python/sglang/srt/model_executor/model_runner_kv_cache_mixin.py]: There is no enable_hisparse parameter in the signature of NPUMLATokenToKVPool.__init__.

File: python/sglang/srt/managers/scheduler_runtime_checker_mixin.py [hunk 2]
```diff
@@ -371,6 +371,11 @@ def self_check_during_idle(self: Scheduler):
                 queue_size += len(self.decode_offload_manager.ongoing_offload)
             if queue_size:
                 return
+        elif self.enable_hisparse:
+            # todo hisparse: resume memory leak check
```
Comment (alphabetc1) [inline python/sglang/srt/managers/scheduler_runtime_checker_mixin.py]: maybe add a `maybe_sleep_on_idle` to avoid spinning?

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -1796,6 +1797,7 @@ def __init__(
         # self.index_k_dtype = torch.float8_e4m3fn
         # self.index_k_scale_dtype = torch.float32
         self.index_head_dim = index_head_dim
+        index_buf_size = index_buf_size if index_buf_size is not None else size
```
Comment (alphabetc1) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```
if index_buf_size is None:
    index_buf_size = size
```

File: python/sglang/srt/mem_cache/hisparse_memory_pool.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    def alloc(self, need_size: int):
+        raise NotImplementedError(
+            "Page size = 1 is not supported in HiSparse allocator"
+        )
+
+    def alloc_device_buffer(self, allocated_indices, need_size: int):
+        assert need_size % self.page_size == 0
+        # clear original reference and isolate the buffer from outside addressing, allocate new buffer if needed
+        hisparse_indices = self.full_to_hisparse_device_index_mapping[allocated_indices]
+        self.full_to_hisparse_device_index_mapping[allocated_indices] = 0
```
Comment (alphabetc1) [inline python/sglang/srt/mem_cache/hisparse_memory_pool.py:187]: It looks like the same logical indices remain visible to prefix reuse via the radix/tree cache, even after HiSparse clears the global logical-to-device KV mapping.

**Sequence:**
1. `cache_unfinished_req()` inserts the request’s prefix into the tree cache, making those logical KV indices reusable by future requests.
2. HiSparse staging then calls `alloc_device_buffer()`, which clears:
   `full_to_hisparse_device_index_mapping[allocated_indices] = 0`.
3. Later, a cache-hit request reuses the same prefix and retrieves those logical indices from the tree cache.
4. During attention, `translate_loc_to_hisparse_device()` translates those still-live logical indices to `0` instead of the actual KV locations.

File: python/sglang/srt/managers/hisparse_coordinator.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+            (max_num_reqs, self.top_k), -1, dtype=torch.int32, device=device
+        )
+        # Scalar tensor: number of real (non-padded) requests in the batch.
+        # Updated before each graph replay so padded blocks early-return.
+        self.num_real_reqs = torch.zeros(1, dtype=torch.int32, device=device)
+
+    def set_decode_producer_stream(self, stream) -> None:
+        self.decode_producer_stream = stream
+
+    def admit_request_into_staging(self, req: Req) -> None:
```
Comment (zoxxxx) [inline python/sglang/srt/managers/hisparse_coordinator.py:125]: Hi @xiezhq-hermann and @hzh0425 ,thanks for your work on this feature! 

Following up on some off-thread discussions with Zhangheng, I've been running some tests with this PR on H200. I encountered an occasional `cudaErrorIllegalAddress` crash. After some debugging, I found it's caused by a race condition in the `admit_request_into_staging` function.

In short, if the scheduler is already executing Batch N+1's forward pass on the `decode_producer_stream`, it will lead to a race and trigger the crash. Adding a specific stream wait appears to resolve the crash in my testing. Here is the suggested fix:

```suggestion
    def admit_request_into_staging(self, req: Req) -> None:
        if self.decode_producer_stream is not None:
            device_module.current_stream().wait_stream(self.decode_producer_stream)
```

I am also very interested in this PR, and I'd be more than happy to help further optimize it.

File: python/sglang/srt/mem_cache/hisparse_memory_pool.py (hunk 4 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/mem_cache/hisparse_memory_pool.py:187]: yes, for now hisparse is intended to be used with --disable-radix-cache

File: python/sglang/srt/managers/hisparse_coordinator.py (hunk 5 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/managers/hisparse_coordinator.py:125]: A decoding step should not happen till the staging finished and formed a new batch

Comment (zoxxxx) [reply python/sglang/srt/managers/hisparse_coordinator.py:125]: I agree that a staged request should not be admitted into a new decode batch before staging finishes. But what I mean is that under `event_loop_overlap`, Batch N+1 forward can already be running on stream before Batch N output processing calls `admit_request_into_staging()`. In that case, offloading and decode forward overlap unless we add an explicit wait_stream dependency.

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def create(
             input_ids = torch.zeros((max_num_token,), dtype=torch.int64)
             input_embeds = torch.zeros((max_num_token, hidden_size), dtype=dtype)
             req_pool_indices = torch.zeros((max_bs,), dtype=torch.int64)
-            seq_lens = torch.full((max_bs,), seq_len_fill_value, dtype=torch.int32)
+            seq_lens = torch.full((max_bs,), seq_len_fill_value, dtype=torch.int64)
```
Comment (xiezhq-hermann) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: @DarkSharpness can you take a look at this change?

File: python/sglang/srt/managers/schedule_batch.py [hunk 7]
```diff
@@ -799,6 +800,9 @@ def __init__(
         # For diffusion LLM
         self.init_diffusion_llm(dllm_config)
 
+        # For hisparse
+        self.staging = False
```
Comment (ispobock) [inline python/sglang/srt/managers/schedule_batch.py:804]: The 'staging' naming is very common. It's already used in dllm and may have some confusion. It's better to have a more specific naming for hisparse.

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -5057,13 +5058,17 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+        parser.add_argument(
+            "--hisparse-config",
             type=str,
-            default=ServerArgs.hierarchical_sparse_attention_extra_config,
+            default=ServerArgs.hisparse_config,
             help="A dictionary in JSON string format for hierarchical sparse attention configuration. "
-            "Required fields: algorithm (str), backend (str). "
-            "All other fields are algorithm-specific and passed to the algorithm constructor. "
-            'Example: \'{"algorithm": "quest", "backend": "flashattention", "sparsity_ratio": 0.7, "min_sparse_prompt_len": 2048}\'',
+            'Example: \'{"top_k": 2048, "device_buffer_size": 4096}\'',
```
Comment (ispobock) [inline python/sglang/srt/server_args.py:5071]: There are serval config here. Maybe we need to have a more detailed usage guide for user later.

File: python/sglang/srt/managers/hisparse_coordinator.py [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+logger = logging.getLogger(__name__)
+
+
+class HiSparseAct(NamedTuple):
+    start_event: device_module.Event
+    finish_event: device_module.Event
+    req: Req
+
+
+class HiSparseCoordinator:
```
Comment (ispobock) [inline python/sglang/srt/managers/hisparse_coordinator.py:30]: Many component logics in this PR should have unit/e2e test coverage ([ref](https://github.com/sgl-project/sglang/blob/main/.claude/skills/write-sglang-test/SKILL.md)). We can add them in followup PR.

File: python/sglang/srt/managers/scheduler.py [hunk 10]
```diff
@@ -2054,30 +2101,40 @@ def get_next_batch_to_run(self) -> Optional[ScheduleBatch]:
...
-            last_bs = self.last_batch.batch_size()
-            self.last_batch.filter_batch(
-                chunked_req_to_exclude=list(chunked_req_to_exclude)
-            )
-            if self.last_batch.batch_size() < last_bs:
-                self.running_batch.batch_is_full = False
-
-            # Merge the new batch into the running batch.
-            if not self.last_batch.is_empty():
+        if self.enable_hisparse:
```
Comment (ispobock) [inline python/sglang/srt/managers/scheduler.py:2104]: Here we have a different branch with the original handling logic. It should be addtitional maintained when later refactor/changes in original branch. Do we have plan to merge it in the original branch with incremental change?

cc: @hnyls2002 for double check

**Review by hnyls2002 — APPROVED** [review]

File: python/sglang/srt/managers/scheduler.py (hunk 10 above)
Comment (hnyls2002) [reply python/sglang/srt/managers/scheduler.py:2104]: I think for all non-hisparse paths, there are no breaks.

**Review by hnyls2002 — APPROVED** [review]
